### PR TITLE
handle DST appropriately in Timestamp.replace

### LINKED ISSUE #18319: Timestamp.replace should handle DST transition gracefully
In fairness, datetime doesn't handle this any better, and differing from that behavior could be considered a drawback.

```
>>> ts = pd.Timestamp('1929-05-05', tz='US/Eastern')
>>> res = ts.replace(month=4)
>>> expected = pd.Timestamp('1929-04-05', tz='US/Eastern')
>>> res == expected
False
>>> expected.tzinfo.normalize(expected) == expected
True
>>> res.tzinfo.normalize(res) == res
False
>>> res.tzinfo.normalize(res)
Timestamp('1929-04-04 22:00:00-0500', tz='US/Eastern')
```

Is this a feature and not a bug for some reason I'm missing?

### PR DESCRIPTION
**Pull request by jbrockmendel**
```
ts = pd.Timestamp('2017-12-03 16:03:24', tz='US/Eastern')
# Timestamp('2017-12-03 16:03:24-0500', tz='US/Eastern')

ts2 = ts.replace(month=6)           # <-- across DST boundary
# master --> Timestamp('2017-06-03 16:03:24-0500', tz='US/Eastern')
# PR     --> Timestamp('2017-06-03 16:03:24-0400', tz='US/Eastern')

ts3 = ts2.tzinfo.normalize(ts2)
# master --> Timestamp('2017-06-03 18:03:24-0400', tz='US/Eastern')
# PR     --> Timestamp('2017-06-03 16:03:24-0400', tz='US/Eastern')

assert ts3 == ts2
assert ts3.hour == ts.hour
```

Both assertions fail under master, pass under this PR.

The whatsnew entry is not super-informative.  Any ideas for a one-sentence description of this fix?

- [x] closes #18319

xref #7825 (I think)
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by gfyoung** [comment]
> closes #18319, closes #7825 (I think)

Any reason why you're not sure about #7825.  If you can run the example code without breaking, that would be considered sufficient to close.

File: doc/source/whatsnew/v0.22.0.txt [hunk 1]
```diff
@@ -248,4 +248,4 @@ Other
 - Fixed a bug where creating a Series from an array that contains both tz-naive and tz-aware values will result in a Series whose dtype is tz-aware instead of object (:issue:`16406`)
 - Fixed construction of a :class:`Series` from a ``dict`` containing ``NaN`` as key (:issue:`18480`)
 - Adding a ``Period`` object to a ``datetime`` or ``Timestamp`` object will now correctly raise a ``TypeError`` (:issue:`17983`)
--
+- :func:`Timestamp.replace` will now handle Daylight Savings transitions gracefully (:issue:`18319`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v0.22.0.txt]: How about "will no longer crash when handling DST transitions" ?

Comment (jreback) [reply doc/source/whatsnew/v0.22.0.txt]: crash is a not an appropriate term for user notes.

File: pandas/tests/scalar/test_timestamp.py [hunk 2]
```diff
@@ -1136,6 +1136,26 @@ def test_timestamp(self):
                 dt = ts.to_pydatetime()
                 assert dt.timestamp() == ts.timestamp()
 
+    def test_replace(self):
```
Comment (jreback) [inline pandas/tests/scalar/test_timestamp.py]: this can't be the right place, we have lots of replace tests

File: pandas/_libs/tslibs/timestamps.pyx [hunk 3]
```diff
@@ -927,8 +927,17 @@ class Timestamp(_Timestamp):
             _tzinfo = tzinfo
 
         # reconstruct & check bounds
-        ts_input = datetime(dts.year, dts.month, dts.day, dts.hour, dts.min,
-                            dts.sec, dts.us, tzinfo=_tzinfo)
+        if _tzinfo is not None and treat_tz_as_pytz(_tzinfo):
+            # be careful about DST transition, #18319
```
Comment (jreback) [inline pandas/_libs/tslibs/timestamps.pyx]: make an informative comments.

why woudn't you always localize?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
lots of tests in #7825, can you add tests / narrow down what is and is not fixed.

**Comment by jbrockmendel** [comment]
> Any reason why you're not sure about #7825. If you can run the example code without breaking, that would be considered sufficient to close.

@gfyoung as jreback mentioned, that issue covered a lot of ground.  The crash there can be ignored, not a pandas issue.  The OP there gave an example of Timestamp.replace that was incorrect, but did not give the expected output.

File: pandas/_libs/tslibs/timestamps.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timestamps.pyx]: `tzinfo.localize` changing the `tzinfo` object is specific to pytz isn't it?

File: pandas/tests/scalar/test_timestamp.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/scalar/test_timestamp.py]: not in test_timestamp... I'll take another look.

Comment (jbrockmendel) [reply pandas/tests/scalar/test_timestamp.py]: Looks like its in test_timezones.

**Comment by jreback** [comment]
@ischwabacher can you have a look

File: pandas/tests/tseries/test_timezones.py [hunk 4]
```diff
@@ -1229,6 +1229,26 @@ def f():
         dt = Timestamp('2013-11-03 01:59:59.999999-0400', tz='US/Eastern')
         assert dt.tz_localize(None) == dt.replace(tzinfo=None)
 
+    def test_replace_across_dst(self):
```
Comment (jreback) [inline pandas/tests/tseries/test_timezones.py]: can you tests with dateutil as well

File: pandas/tests/tseries/test_timezones.py [hunk 5]
```diff
@@ -1229,6 +1229,26 @@ def f():
...
+        # GH#18319
```
Comment (jreback) [inline pandas/tests/tseries/test_timezones.py]: can you provide a 1-liner about what this is testing (I know the title is informative, but a little color)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/tseries/test_timezones.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/tests/tseries/test_timezones.py]: Sure

File: pandas/tests/tseries/test_timezones.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/tseries/test_timezones.py]: The test is pretty specific to pytz API

**Comment by jreback** [comment]
needs a rebase. also pls add all of the examples from #7825 ; if something doesn't work, then you can open new issue for that part.

**Comment by jreback** [comment]
further there are a number of timezone issues that this (and/or other fixes) might have resolved. should investigate. anyone want to take this on? @gfyoung @chris-b1 ?

**Comment by jbrockmendel** [comment]
> needs a rebase.

Will do.

> also pls add all of the examples from #7825 ; if something doesn't work, then you can open new issue for that part.

#7825 is all over the place (I probably shouldn't have referenced it above).  The example in the OP I _think_ is handled, but the poster didn't specify the expected behavior.  I think the thing to do is keep that open and ask the OP there to verify if the new behavior closes that issue.

File: doc/source/whatsnew/v0.22.0.txt [hunk 6]
```diff
@@ -320,5 +320,10 @@ Categorical
 Other
 ^^^^^
 
+- Improved error message when attempting to use a Python keyword as an identifier in a numexpr query (:issue:`18221`)
+- Fixed a bug where creating a Series from an array that contains both tz-naive and tz-aware values will result in a Series whose dtype is tz-aware instead of object (:issue:`16406`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: you have rebase issues here

**Review by jreback — CHANGES_REQUESTED** [review]
I would like to include some tests on the original issue #7825 if they are not working xfail them.

File: doc/source/whatsnew/v0.22.0.txt (hunk 6 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v0.22.0.txt]: Will fix.  Not a big deal, but for changes limited to whatsnew is skipping the CI a viable option?  The CI turnaround time is close to a full day and that file gets touched a lot.

**Comment by jbrockmendel** [comment]
> I would like to include some tests on the original issue #7825 if they are not working xfail them.

I'll take another look, but based on the last two looks I concluded that this is not actionable until the author there provides the expected output for the example cases.

Simpler to treat #18319 as the "original issue".

**Comment by jbrockmendel** [comment]
There's a follow-up to this waiting in the wings that fixes a couple of related bugs.

File: pandas/tests/tseries/test_timezones.py (hunk 4 above)
Comment (pganssle) [reply pandas/tests/tseries/test_timezones.py]: The test is specific to the `pytz` API because the *problem* is specific to the `pytz` API, so you can't write tests that currently fail with `dateutil`, *but* you can, in fact, write this test to be agnostic as to whether the test takes a `pytz` or `dateutil` zone.

Rather than using `localize` and `normalize`, express the "expected" datetimes as UTC datetimes and use `.astimezone`. `pytz` and `dateutil` support `astimezone` (and, in fact, `pytz.normalize` essentially just uses `.astimezone` under the hood). Since `pandas` provides their own API layer on top of `datetime`, it is *also* agnostic to the timezone provider.

**Comment by pganssle** [comment]
I think this needs at least two additional tests - 1. a test for when `datetime.replace` creates an ambiguous time and 2. a test for when `datetime.replace` creates an imaginary time.

I *think* that the way it's done right now, `pytz` and `dateutil` zones will do different things in these situations.

File: pandas/tests/tseries/test_timezones.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/tseries/test_timezones.py]: > Rather than using localize and normalize, express the "expected" datetimes as UTC datetimes and use .astimezone

Is this a suggestion for this test or more generally?

**Comment by jbrockmendel** [comment]
> I think this needs at least two additional tests - 1. a test for when datetime.replace creates an ambiguous time and 2. a test for when datetime.replace creates an imaginary time.

Good idea, will do.

File: pandas/tests/tseries/test_timezones.py (hunk 4 above)
Comment (pganssle) [reply pandas/tests/tseries/test_timezones.py]: > Is this a suggestion for this test or more generally?

In this test at least, since with tests you can always engineer your datetime literals such that you never have to use anything except `astimezone`.

It's also just a useful piece of information to have, that `astimezone` is one of the few timezone functions where `pytz` behaves more or less nicely. You only need `normalize` semantics with `pytz` for things like `replace` or "calendar" arithmetic, where you want to change the naive portion of the datetime to a specific value and then change the time zone offset to match. If you want "absolute" arithmetic (where you want to go forward a certain number of hours or seconds or something), then `the_operation(dt.astimezone(UTC)).astimezone(dt.tzinfo)` will always give the right answer.

In this case you know what the correct answer *should* be, in absolute time, so you can just declare your initial variable as the correct answer in `UTC` and convert that to the time zone you care about. If you do it generically like that, you are insulated from any weird quirks of `pytz`'s interface, and you get `dateutil` support for free (so you can parametrize this test using `pytz` and `dateutil` zones).

Of course, this method will fail if you try to construct an imaginary time, since there is no mapping between UTC and imaginary times.

**Comment by jreback** [comment]
I bet this will fix #18785 as well.

**Comment by jreback** [comment]
I added an example for #18785

**Comment by jbrockmendel** [comment]
> I think this needs at least two additional tests - 1. a test for when datetime.replace creates an ambiguous time and 2. a test for when datetime.replace creates an imaginary time.
>
> I think that the way it's done right now, pytz and dateutil zones will do different things in these situations.

I've added these tests and you're right, these do different things.  1AM on 2014-11-05 is ambiguous in US/Eastern:

```
>>> ts_dateutil = pd.Timestamp('2014-11-05 01:00:00', tz=dateutil.tz.gettz('US/Eastern'))
>>> ts_pytz = pd.Timestamp('2014-11-05 01:00:00', tz=pytz.timezone('US/Eastern'))
>>> ts_dateutil == ts_pytz
True
>>> res_dateutil = ts_dateutil.replace(day=2)
>>> res_dateutil = Timestamp('2014-11-02 01:00:00-0400', tz='dateutil//usr/share/zoneinfo/US/Eastern')
>>> res_pytz = ts_pytz.replace(day=2)
>>> res_pytz
Timestamp('2014-11-02 01:00:00-0500', tz='US/Eastern')
```

So the pytz version resolves to the later of the two possibilities while the dateutil version resolves to the earlier.  It isn't clear to me that there is an unambiguous "right" thing to do in this case, so as far as pandas is concerned just deferring to pytz/dateutil may be the way to go.  Thoughts?

As for imaginary times, both pytz and dateutil versions (both in the PR and in master) fail to raise when `ts.replace()` results in a non-existent time.  My preference would be to address this in a separate issue/PR.

**Comment by pganssle** [comment]
> So the pytz version resolves to the later of the two possibilities while the dateutil version resolves to the earlier. It isn't clear to me that there is an unambiguous "right" thing to do in this case, so as far as pandas is concerned just deferring to pytz/dateutil may be the way to go. Thoughts?

There is indeed an unambiguous right way to do things. The Python documentation before 3.6 suggested that ambiguous times should always resolve to the standard time (the later time). As of Python 3.6, unfortunately this is inverted because the default value of `fold` is 0 and `fold=1` is interpreted as the *second* time. So dateutil is correct in this case, but you are welcome to defer this as "`pytz`'s problem".

> As for imaginary times, both pytz and dateutil versions (both in the PR and in master) fail to raise when `ts.replace()` results in a non-existent time. My preference would be to address this in a separate issue/PR.

No time zone hooks are called on `replace` or on any constructor. If you want to retain the semantic meaning of `replace`, then do nothing. If you want `replace` to always produce a valid date, it's necessary to choose a default behavior.

To me, it seems like this PR is mostly about making `pytz` act like `dateutil`, since `dateutil` implements the pythonic `tzinfo` interface and `pytz` uses its own interface. If you think of it this way, then you should basically just be supporting PEP 495 fold for ambiguous times (not difficult to backport when you're already wrapping `datetime` internally) and allowing `replace` to create imaginary times (though if you want to enforce real datetimes in a separate PR, or add some mechanism to normalize imaginary datetimes, that's reasonable).

File: pandas/tests/tseries/test_timezones.py [hunk 7]
```diff
@@ -1228,6 +1228,27 @@ def f():
...
+        # GH#18319 check that 1) timezone is correctly normalized and
+        # 2) that hour is not incorrectly changed by this normalization
+        tz = pytz.timezone('US/Eastern')
+
+        ts_naive = Timestamp('2017-12-03 16:03:30')
+        ts_aware = tz.localize(ts_naive)
+
+        # Preliminary sanity-check
+        assert ts_aware == ts_aware.tzinfo.normalize(ts_aware)
+
```
Comment (jreback) [inline pandas/tests/tseries/test_timezones.py]: need to parametrize this across dateutil & pytz, so you need to move this to ``TestTimeZoneSupportPytz`` and use ``self.tz(...)``

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
did you cover @pganssle requests for test?

**Comment by jbrockmendel** [comment]
> did you cover pganssle requests for test?

@pganssle correct me if I'm wrong but I think the consensus was that there are some lower-level changes that might avoid this class of problem longer-term, but it was OK to consider that out of scope for this PR.  (Possibly what you had in mind for #18595?).

**Comment by pganssle** [comment]
@jreback While #18595 and this issue share the same root cause - namely that `pytz` uses a different mechanism for assigning time zones than what is expected in the `datetime` library - the solutions are different. #18595 is just about having a canonical time zone object for a given series, whereas this issue is about the fact that after every operation with a datetime with attached `pytz` zone, the zone offsets need to be re-normalized.

To put it another way, if you think of pytz as having a "parent" zone (the one you get by calling `pytz.timezone`) and "child" zones (the ones that are attached to the `datetime` by `.localize` and `.normalize`, #18595 is about having a standard way to retrieve the "parent" zone, while this PR is about retrieving the correct "child" zone for a given datetime.

**Comment by pganssle** [comment]
This is a sort of "out there" idea, but there is one (possibly not particularly performant) thing you could do that would unify the interface between `pytz` and `dateutil`, and also solve #18595, which is to write a wrapper around `pytz` that just has the standard `tzinfo` interface. A rough sketch (eliding some of the details like how to handle ambiguous and imaginary times):

```python
from datetime import tzinfo
from datetime import datetime, timedelta

import functools

import pytz

def _localize(f):
    @functools.wraps(f)
    def inner_func(self, dt):
        return f(self, self._zone.localize(dt.replace(tzinfo=None)))

    return inner_func

class pytz_zonewrapper(tzinfo):
    def __init__(self, pytz_zone):
        self._zone = pytz.timezone(str(pytz_zone))

    @_localize
    def utcoffset(self, dt):
        return dt.utcoffset()

    @_localize
    def tzname(self, dt):
        return dt.tzname()

    @_localize
    def dst(self, dt):
        return dt.dst()
```

I would say for now go with @jbrockmendel's approach, but I would think that it might be nice to remove all the pytz-specific code cluttering up time zone handling.

Another option might be to just internally start using `dateutil` zones. I don't think it's *guaranteed*, but from my testing `tz.gettz(str(pytz_zone))` will always give you the appropriate, valid `dateutil` zone. I believe that `pytz` is still considerably faster than `dateutil` in many areas, which I'm hoping to take care of at some point soon but haven't gotten around to yet.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks, I left #7825 open until we resolve whether this closes it.
